File: app/src/main/java/com/colourizmus/model/CustomColour.java
package com.colourizmus.model;

import android.arch.persistence.room.ColumnInfo;
import android.arch.persistence.room.Entity;
import android.arch.persistence.room.Ignore;
import android.arch.persistence.room.Index;
import android.arch.persistence.room.PrimaryKey;
import android.support.annotation.ColorInt;

/**
 * The CustomColour represents a single, named colour that can be marked as favourite and is persisted in on-disk database.
 */
@Entity ( tableName = CustomColour.TABLE, indices = @Index(value = CustomColour.COLUMN_NAME, unique = true))
public class CustomColour {

    @Ignore public static final String TABLE = "colour";
    @Ignore public static final String COLUMN_PK = "value";
    @Ignore public static final String COLUMN_NAME = "name";
    @Ignore public static final String COLUMN_IS_FAVOURITE = "is_favourite";

    @PrimaryKey (autoGenerate = false)
    @ColumnInfo (name = COLUMN_PK)
    private final @ColorInt int value;

    @ColumnInfo (name = COLUMN_NAME)
    private String name;

    @ColumnInfo (name = COLUMN_IS_FAVOURITE)
    private boolean isFavourite;

    protected CustomColour(String name, @ColorInt int value) {
        //TODO !! VALIDATION !!

        this.name = name;
        this.value = value;
        this.isFavourite = false;
    }


    @ColorInt
    protected int getValue() {
        return value;
    }

    protected String getName() {
        return name;
    }

    protected void setName(String name) {
        if (name != null) //TODO - check for uniqueness?
            this.name = name;
    }

    protected boolean getIsFavourite() {
        return isFavourite;
    }

    protected void setIsFavourite(boolean isFavourite) {
        this.isFavourite = isFavourite;
    }
}

File: app/src/main/java/bg/o/sim/colourizmus/model/CustomColour.kt
package bg.o.sim.colourizmus.model

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey
import androidx.annotation.ColorInt
import java.io.Serializable

const val TABLE: String = "colour"
const val COLUMN_PK: String = "value"
const val COLUMN_NAME: String = "name"
const val COLUMN_IS_FAVOURITE: String = "is_favourite"

/**
 * The [CustomColour] represents a single, named, colour that
 * can be marked as favourite and is persisted in on-disk database.
 */
@Entity(tableName = TABLE, indices = [(Index(value = [COLUMN_NAME], unique = true))])
class CustomColour(@ColorInt value: Int, name: String) : Serializable {

    @PrimaryKey(autoGenerate = false)
    @ColumnInfo(name = COLUMN_PK)
    @ColorInt val value: Int = value

    @ColumnInfo(name = COLUMN_NAME)
    var name: String = name

    @ColumnInfo(name = COLUMN_IS_FAVOURITE)
    var isFavourite: Boolean = false

    // TODO: 16/02/18 - include name in hash? Yes? No?
    override fun hashCode(): Int = value

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (this.javaClass != other?.javaClass) return false

        other as CustomColour

        if (value != other.value) return false
        if (name != other.name) return false

        return true
    }
}
